torch/_vmap_internals.py

```diff
@@ -61,6 +93,13 @@ def _validate_outputs(outputs, fn_name):
             continue
         raise ValueError(ELEMENT_MUST_BE_TENSOR.format(fn=fn_name, out=type(output), idx=idx))
 
+def _check_out_dims_is_int_or_int_tuple(out_dims, fn_name):
+    if isinstance(out_dims, int):
+        return
+    if not isinstance(out_dims, tuple) or \
+            not all([isinstance(out_dim, int) for out_dim in out_dims]):
+        raise ValueError(OUT_DIMS_MUST_BE_INT_OR_TUPLE_OF_INT.format(out_dims=out_dims, fn=fn_name))
```

Comment:
Now I am wondering, why the argument validation isn't done in the C binding code. A lot of this would be handled automatically (and more efficiently) by our existing Python binding code. I guess because there isn't a convenient "API" you can expose from C++ (but I wonder if this necessarily has to be the case...)

Reply:
Right now we accept `out_dims` as a tuple of ints. In the future, we'll want to accept `out_dims` as an arbitrary python nested data structure. It can have dictionaries, tuples, lists, etc nested together. Our C++ arg parser cannot handle something like that so we do it at the Python level as an initial implementation.

Furthermore, our C++ arg parser doesn't differentiate between "lists of int" and "tuples of int" (we only want to support tuple of int for now). There's a semantic difference between those two for vmap. `out_dims` should be a data structure that mirrors the data structure of the output of the function. In python, if a function returns multiple outputs, then it returns a tuple with said outputs. `out_dims` with a list of ints should be used in the case where a function returns a list of tensors.